### ISSUE TITLE
Protect npm publish when docker tests are pending

### ISSUE THREAD

**Issue by fisehara** [product-os/flowzone#195]
When docker tests are executed but the npm tests have already succeeded the npm publish is starting.

As there is no conditional needs, the npm publish has to wait for the docker test job always and needs to evaluated further.

https://stackoverflow.com/questions/65844033/conditional-needs-in-github-action
copy-pasta from stackoverflow: 
``` yaml
jobs:
  init_job:
    runs-on: ubuntu-latest
    steps:
      - run: echo "initial run"

  conditional_job:
    runs-on: ubuntu-latest
    needs: init_job
    if: ${{ your condition here }}
    steps:
      - run: echo "conditonal run"

  final_job:
    needs: conditional_job
    if: always()
    steps:
      - name: fail if conditional job failed
        if: ${{ needs.conditional_job.result == 'failure' }}
        run: exit 1
      - run: echo "final run"
```

**Comment by klutchell** [comment]
We are using this type of logic in the `custom_publish` job to wait for other test jobs, but still run if a dependency is skipped.
https://github.com/product-os/flowzone/blob/4e01ead9ffd98ad6415fdc4b1d94cdaed35cac2c/.github/workflows/flowzone.yml#L1041-L1056

This was not put in place for `publish_npm` as it was unclear if there was a use case for npm publish to wait for docker tests, or docker publish to wait for npm tests. Though it may be more clear if all publish jobs follow the same logic and only require 'success' from their upstream project-type test.

Be warned, this conditional logic gets really hard to follow if we use it too often, so use sparingly to avoid spaghetti code!

**Comment by klutchell** [comment]
Another option I briefly explored was to have `all_tests` and `all_published` jobs using the types of conditions above to break the workflow into "stages". In this case `all_tests` would pass only if all dependencies were success or skipped, and no publish jobs would run without `all_test` being success.

The other advantage here is it becomes logical to add `all_tests` and `all_published` to the branch protection rules.
